src/state.rs

```diff
@@ -259,6 +267,206 @@ impl DemandProxyState {
         false
     }
 
+    // this should only be called once per request (for the workload itself and potentially its waypoint)
+    pub async fn load_balance(
+        &self,
+        dst_workload: &Workload,
+        src_workload: &Workload,
+        dns_resolver_config: &ResolverConfig,
+        dns_resolver_opts: &ResolverOpts,
+        metrics: Arc<proxy::Metrics>,
+    ) -> Result<IpAddr, Error> {
+        // TODO: add more sophisticated routing logic, perhaps based on ipv4/ipv6 support underneath us.
+        // if/when we support that, this function may need to move to get access to the necessary metadata.
+        // Randomly pick an IP
+        // TODO: do this more efficiently, and not just randomly
+        if let Some(ip) = dst_workload.workload_ips.choose(&mut rand::thread_rng()) {
+            return Ok(*ip);
+        }
+        if dst_workload.hostname.is_empty() {
+            debug!(
+                "workload {} has no suitable workload IPs for routing",
+                dst_workload.name
+            );
+            return Err(Error::NoValidDestination(Box::new(dst_workload.clone())));
+        }
+        let ip = self
+            .load_balance_for_workload(
+                dst_workload,
+                src_workload,
+                dns_resolver_config,
+                dns_resolver_opts,
+                metrics,
+            )
+            .await?;
+        Ok(ip)
+    }
+
+    async fn load_balance_for_workload(
```

Comment:
should make it more readable, it is resolving hostname

Reply:
addressed in https://github.com/istio/ztunnel/pull/612/commits/c8615127acf2d41d00812688db412f1d69d514fc